pkg/prometheus/agent/daemonset.go

```diff
@@ -0,0 +1,319 @@
+	// We set some defaults if some fields are not present, and we want those fields set in the original Prometheus object before building the DaemonSetSpec.
+	p.SetCommonPrometheusFields(cpf)
+	spec, err := makeDaemonSetSpec(p, config, cg, tlsSecrets)
+	if err != nil {
+		return nil, fmt.Errorf("make DaemonSet spec: %w", err)
+	}
+
+	// do not transfer kubectl annotations to the daemonset so it is not
+	// pruned by kubectl
+	annotations := make(map[string]string, 0)
+	for key, value := range objMeta.GetAnnotations() {
+		if !strings.HasPrefix(key, "kubectl.kubernetes.io/") {
+			annotations[key] = value
+		}
+	}
+	daemonSet := &appsv1.DaemonSet{Spec: *spec}
+
+	operator.UpdateObject(
+		daemonSet,
+		operator.WithName(name),
+		operator.WithAnnotations(annotations),
+		operator.WithAnnotations(config.Annotations),
+		operator.WithLabels(objMeta.GetLabels()),
+		operator.WithLabels(map[string]string{
+			prompkg.PrometheusNameLabelName: objMeta.GetName(),
+			prompkg.PrometheusModeLabeLName: prometheusMode,
+		}),
+		operator.WithLabels(config.Labels),
+		operator.WithManagingOwner(p),
+	)
+
+	if cpf.ImagePullSecrets != nil && len(cpf.ImagePullSecrets) > 0 {
+		daemonSet.Spec.Template.Spec.ImagePullSecrets = cpf.ImagePullSecrets
+	}
+
+	if cpf.HostNetwork {
+		daemonSet.Spec.Template.Spec.DNSPolicy = v1.DNSClusterFirstWithHostNet
+	}
+
+	return daemonSet, nil
+}
+
+func makeDaemonSetSpec(
+	p monitoringv1.PrometheusInterface,
+	c *prompkg.Config,
+	cg *prompkg.ConfigGenerator,
+	tlsSecrets *operator.ShardedSecret,
+) (*appsv1.DaemonSetSpec, error) {
+	cpf := p.GetCommonPrometheusFields()
+
+	pImagePath, err := operator.BuildImagePath(
+		operator.StringPtrValOrDefault(cpf.Image, ""),
+		operator.StringValOrDefault("", c.PrometheusDefaultBaseImage),
+		operator.StringValOrDefault(cpf.Version, operator.DefaultPrometheusVersion),
+		"",
+		"",
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	cpf.EnableFeatures = append(cpf.EnableFeatures, "agent")
+	promArgs := prompkg.BuildCommonPrometheusArgs(cpf, cg)
+	promArgs = appendAgentArgs(promArgs, cg, cpf.WALCompression)
+
+	var ports []v1.ContainerPort
+	if !cpf.ListenLocal {
+		ports = []v1.ContainerPort{
+			{
+				Name:          cpf.PortName,
+				ContainerPort: 9090,
+				Protocol:      v1.ProtocolTCP,
+			},
+		}
+	}
+
+	volumes, promVolumeMounts, err := prompkg.BuildCommonVolumes(p, tlsSecrets)
+	if err != nil {
+		return nil, err
+	}
+
+	configReloaderVolumeMounts := []v1.VolumeMount{
+		{
+			Name:      "config",
+			MountPath: prompkg.ConfDir,
+		},
+		{
+			Name:      "config-out",
+			MountPath: prompkg.ConfOutDir,
+		},
+	}
+
+	var configReloaderWebConfigFile string
+
+	// Mount web config and web TLS credentials as volumes.
+	// We always mount the web config file for versions greater than 2.24.0.
+	// With this we avoid redeploying prometheus when reconfiguring between
+	// HTTP and HTTPS and vice-versa.
```

Comment:
I'd simplify and just consider that users must use a recent Prometheus version for DaemonSet deployment.

Reply:
Yeah, it's a new mode so we don't need this check

Reply:
Now I'm not entirely sure yet we should remove this check in DaemonSet. If we remove this check, it means we will officially require users to use >=2.24 right? If that's the case, we can remove this check and should clearly write the requirement in the doc.

Reply:
Let me try to clarify: I would declare that Daemonset mode requires a "recent" version of Prometheus (for instance at least 2.45.0 which the current long-term support version) so we can simplify the support matrix. The check should be done as early as possible so makeDaemonSet() can safely assume a minimal version.

Reply:
Yeah I wanted to clarify this, because we had a discussion in Slack (with @ArthurSens too) about whether we should limit the Prometheus version in DaemonSet: https://cloud-native.slack.com/archives/C071STDT4FP/p1718703164200639

If we agree to require a specific recent version, I'll remove this check then. I'll also note so that we'll remember to clarify this is the doc of DaemonSet.